File: Classes/View Controllers/Tabs/Artists/ArtistsViewController.m
//
//  ArtistsViewController.m
//  iSub
//
//  Created by Benjamin Baron on 12/22/20.
//  Copyright © 2020 Ben Baron. All rights reserved.
//

#import "ArtistsViewController.h"
#import "ServerListViewController.h"
#import "FolderDropdownControl.h"
#import "ViewObjectsSingleton.h"
#import "Defines.h"
#import "Flurry.h"
#import "SavedSettings.h"
#import "MusicSingleton.h"
#import "EX2Kit.h"
#import "Swift.h"
#import <QuartzCore/QuartzCore.h>

@interface ArtistsViewController() <UISearchBarDelegate, APILoaderDelegate, FolderDropdownDelegate>
@end

@implementation ArtistsViewController

#pragma mark - Lifecycle

- (void)createDataModel {
    NSInteger mediaFolderId = settingsS.rootArtistsSelectedFolderId.integerValue;
    self.dataModel = [[RootArtistsViewModel alloc] initWithMediaFolderId:mediaFolderId delegate:self];
}

- (void)viewDidLoad  {
    [super viewDidLoad];
    
    [self createDataModel];
    
    self.title = @"Artists";
    self.view.backgroundColor = [UIColor colorNamed:@"isubBackgroundColor"];
    
    //Set defaults
    self.isSearching = NO;
    self.isCountShowing = NO;
    
    [NSNotificationCenter addObserverOnMainThread:self selector:@selector(serverSwitched) name:ISMSNotification_ServerSwitched];
    [NSNotificationCenter addObserverOnMainThread:self selector:@selector(updateFolders) name:ISMSNotification_ServerCheckPassed];
        
    // Add the pull to refresh view
    __weak ArtistsViewController *weakSelf = self;
    self.tableView.refreshControl = [[RefreshControl alloc] initWithHandler:^{
        [weakSelf loadData:settingsS.rootArtistsSelectedFolderId.integerValue];
    }];
    
    [self.tableView registerClass:BlurredSectionHeader.class forHeaderFooterViewReuseIdentifier:BlurredSectionHeader.reuseId];
    [self.tableView registerClass:UniversalTableViewCell.class forCellReuseIdentifier:UniversalTableViewCell.reuseId];
    self.tableView.rowHeight = Defines.rowHeight;
    
    if (self.dataModel.isCached) {
        [self addCount];
        [self.tableView setContentOffset:CGPointMake(0, 0) animated:NO];
    }
    
    [NSNotificationCenter addObserverOnMainThread:self selector:@selector(addURLRefBackButton) name:UIApplicationDidBecomeActiveNotification];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    [self addURLRefBackButton];
    [self addShowPlayerButton];
    
    if (!viewObjectsS.isArtistsLoading) {
        if (!self.dataModel.isCached) {
            [self loadData:settingsS.rootArtistsSelectedFolderId.integerValue];
        }
    }
    
    [Flurry logEvent:@"FoldersTab"];
}

- (void)dealloc {
    [NSNotificationCenter removeObserverOnMainThread:self];
    self.dataModel.delegate = nil;
    self.dropdown.delegate = nil;
}

#pragma mark Loading

- (void)updateCount {
    if (self.dataModel.count == 1) {
        self.countLabel.text = [NSString stringWithFormat:@"%lu Artist", (unsigned long)self.dataModel.count];
    } else {
        self.countLabel.text = [NSString stringWithFormat:@"%lu Artists", (unsigned long)self.dataModel.count];
    }
    
    NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
    [formatter setDateStyle:NSDateFormatterMediumStyle];
    [formatter setTimeStyle:NSDateFormatterShortStyle];
    self.reloadTimeLabel.text = [NSString stringWithFormat:@"last reload: %@", [formatter stringFromDate:self.dataModel.reloadDate]];
}

- (void)removeCount {
    self.tableView.tableHeaderView = nil;
    self.isCountShowing = NO;
}

- (void)addCount {
    self.isCountShowing = YES;
    
    self.headerView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 320, 157)];
    self.headerView.autoresizingMask = UIViewAutoresizingFlexibleWidth;
    self.headerView.backgroundColor = self.view.backgroundColor;
    
    self.countLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, 9, 320, 30)];
    self.countLabel.autoresizingMask = UIViewAutoresizingFlexibleWidth;
    self.countLabel.textColor = UIColor.labelColor;
    self.countLabel.textAlignment = NSTextAlignmentCenter;
    self.countLabel.font = [UIFont boldSystemFontOfSize:32];
    [self.headerView addSubview:self.countLabel];
    
    self.reloadTimeLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, 40, 320, 14)];
    self.reloadTimeLabel.autoresizingMask = UIViewAutoresizingFlexibleWidth;
    self.reloadTimeLabel.textColor = UIColor.secondaryLabelColor;
    self.reloadTimeLabel.textAlignment = NSTextAlignmentCenter;
    self.reloadTimeLabel.font = [UIFont systemFontOfSize:11];
    [self.headerView addSubview:self.reloadTimeLabel];
    
    self.dropdown = [[FolderDropdownControl alloc] initWithFrame:CGRectMake(50, 61, 220, 40)];
    self.dropdown.delegate = self;
    [self.dropdown selectFolderWithId:self.dataModel.mediaFolderId];
    [self.headerView addSubview:self.dropdown];
    
    self.searchBar = [[UISearchBar  alloc] initWithFrame:CGRectMake(0, 111, 320, 40)];
    self.searchBar.autoresizingMask = UIViewAutoresizingFlexibleWidth;
    self.searchBar.searchBarStyle = UISearchBarStyleMinimal;
    self.searchBar.delegate = self;
    self.searchBar.autocorrectionType = UITextAutocorrectionTypeNo;
    self.searchBar.placeholder = @"Folder name";
    [self.headerView addSubview:self.searchBar];
    
    [self updateCount];
    
    // Special handling for voice over users
    if (UIAccessibilityIsVoiceOverRunning()) {
        // Add a refresh button
        UIButton *voiceOverRefresh = [UIButton buttonWithType:UIButtonTypeCustom];
        voiceOverRefresh.frame = CGRectMake(0, 0, 50, 50);
        [voiceOverRefresh addTarget:self action:@selector(reloadAction:) forControlEvents:UIControlEventTouchUpInside];
        voiceOverRefresh.accessibilityLabel = @"Reload Folders";
        [self.headerView addSubview:voiceOverRefresh];
        
        // Resize the two labels at the top so the refresh button can be pressed
        self.countLabel.frame = CGRectMake(50, 5, 220, 30);
        self.reloadTimeLabel.frame = CGRectMake(50, 36, 220, 12);
    }
    
    self.tableView.tableHeaderView = self.headerView;
}

- (void)cancelLoad {
    [self.dataModel cancelLoad];
    [viewObjectsS hideLoadingScreen];
    [self.tableView.refreshControl endRefreshing];
}

- (void)loadData:(NSInteger)mediaFolderId  {
    [self.dropdown updateFolders];
    viewObjectsS.isArtistsLoading = YES;
    [viewObjectsS showAlbumLoadingScreenOnMainWindowWithSender:self];
    self.dataModel.mediaFolderId = mediaFolderId;
    [self.dataModel startLoad];
}

- (void)loadingFinished:(APILoader *)loader {
    if (self.isCountShowing) {
        [self updateCount];
    } else {
        [self addCount];
    }
    
    [self.tableView reloadData];
    
    viewObjectsS.isArtistsLoading = NO;
    
    // Hide the loading screen
    [viewObjectsS hideLoadingScreen];
    
    [self.tableView.refreshControl endRefreshing];
}

- (void)loadingFailed:(APILoader *)loader error:(NSError *)error {
    viewObjectsS.isArtistsLoading = NO;
    
    // Hide the loading screen
    [viewObjectsS hideLoadingScreen];
    
    [self.tableView.refreshControl endRefreshing];
    
    // Inform the user that the connection failed.
    // NOTE: Must call after a delay or the refresh control won't hide
    [EX2Dispatch runInMainThreadAfterDelay:0.3 block:^{
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Subsonic Error"
                                                                       message:error.localizedDescription
                                                                preferredStyle:UIAlertControllerStyleAlert];
        [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleCancel handler:nil]];
        [self presentViewController:alert animated:YES completion:nil];
    }];
}

#pragma mark Folder Dropdown Delegate

- (void)folderDropdownMoveViewsY:(float)y {
    [self.tableView performBatchUpdates:^{
        self.tableView.tableHeaderView.height += y;
        self.searchBar.y += y;
        self.tableView.tableHeaderView = self.tableView.tableHeaderView;

        NSSet *visibleSections = [NSSet setWithArray:[[self.tableView indexPathsForVisibleRows] valueForKey:@"section"]];
        for (NSNumber *section in visibleSections) {
            UIView *sectionHeader = [self.tableView headerViewForSection:section.intValue];
            sectionHeader.y += y;
        }
    } completion:nil];
}

- (void)folderDropdownSelectFolder:(NSInteger)mediaFolderId {
    // Save the default
    settingsS.rootArtistsSelectedFolderId = @(mediaFolderId);
    
    // Reload the data
    self.dataModel.mediaFolderId = mediaFolderId;
    self.isSearching = NO;
    if (self.dataModel.isCached) {
        [self.tableView reloadData];
        [self updateCount];
    } else {
        [self loadData:mediaFolderId];
    }
}

- (void)serverSwitched {
    [self createDataModel];
    if (!self.dataModel.isCached) {
        [self.tableView reloadData];
        [self removeCount];
    }
    [self folderDropdownSelectFolder:-1];
}

- (void)updateFolders {
    [self.dropdown updateFolders];
}

#pragma mark Button handling methods

- (void)reloadAction:(id)sender {
    [self loadData:settingsS.rootArtistsSelectedFolderId.integerValue];
}

#pragma mark SearchBar

- (void)searchBarTextDidBeginEditing:(UISearchBar *)theSearchBar {
    if (self.isSearching) return;
    
    self.isSearching = YES;
    [self.dataModel clearSearch];
    
    [self.dropdown closeDropdownFast];
    [self.tableView setContentOffset:CGPointMake(0, 104) animated:YES];
    
    if (theSearchBar.text.length == 0) {
        [self createSearchOverlay];
    }
    
    // Add the done button.
    self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemDone target:self action:@selector(searchBarSearchButtonClicked:)];
}

- (void)searchBarTextDidEndEditing:(UISearchBar *)searchBar {
    [searchBar resignFirstResponder];
}

- (void)searchBar:(UISearchBar *)theSearchBar textDidChange:(NSString *)searchText {
    [self.dataModel clearSearch];
    if (searchText.length > 0)  {
        [self hideSearchOverlay];
        [self.dataModel searchWithName:self.searchBar.text];
    } else {
        [self createSearchOverlay];
        [self.tableView setContentOffset:CGPointMake(0, 104) animated:NO];
    }
    [self.tableView reloadData];
}

- (void)searchBarSearchButtonClicked:(UISearchBar *)theSearchBar {
    [self updateCount];
    
    self.searchBar.text = @"";
    [self.searchBar resignFirstResponder];
    [self hideSearchOverlay];
    self.isSearching = NO;
    
    self.navigationItem.leftBarButtonItem = nil;
    [self.dataModel clearSearch];
    [self.tableView reloadData];
    [self.tableView setContentOffset:CGPointMake(0, 104) animated:YES];
}

- (void)createSearchOverlay {
    UIBlurEffectStyle effectStyle = self.traitCollection.userInterfaceStyle == UIUserInterfaceStyleDark ? UIBlurEffectStyleSystemUltraThinMaterialLight : UIBlurEffectStyleSystemUltraThinMaterialDark;
    self.searchOverlay = [[UIVisualEffectView alloc] initWithEffect:[UIBlurEffect effectWithStyle:effectStyle]];
    self.searchOverlay.translatesAutoresizingMaskIntoConstraints = NO;
    
    UIButton *dismissButton = [UIButton buttonWithType:UIButtonTypeCustom];
    dismissButton.translatesAutoresizingMaskIntoConstraints = NO;
    [dismissButton addTarget:self action:@selector(searchBarSearchButtonClicked:) forControlEvents:UIControlEventTouchUpInside];
    [self.searchOverlay.contentView addSubview:dismissButton];
    
    [self.view addSubview:self.searchOverlay];
    
    [NSLayoutConstraint activateConstraints:@[
        [self.searchOverlay.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor],
        [self.searchOverlay.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor],
        [self.searchOverlay.topAnchor constraintEqualToAnchor:self.view.topAnchor constant:50],
        [self.searchOverlay.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor],
        
        [dismissButton.leadingAnchor constraintEqualToAnchor:self.searchOverlay.leadingAnchor],
        [dismissButton.trailingAnchor constraintEqualToAnchor:self.searchOverlay.trailingAnchor],
        [dismissButton.topAnchor constraintEqualToAnchor:self.searchOverlay.topAnchor],
        [dismissButton.bottomAnchor constraintEqualToAnchor:self.searchOverlay.bottomAnchor]
    ]];
    
    // Animate the search overlay on screen
    self.searchOverlay.alpha = 0.0;
    [UIView animateWithDuration:0.2 delay:0 options:UIViewAnimationOptionCurveEaseOut animations:^{
        self.searchOverlay.alpha = 1;
    } completion:nil];
}

- (void)hideSearchOverlay {
    if (self.searchOverlay) {
        // Animate the search overlay off screen
        [UIView animateWithDuration:0.2 delay:0 options:UIViewAnimationOptionCurveEaseOut animations:^{
            self.searchOverlay.alpha = 0;
        } completion:^(BOOL finished) {
            [self.searchOverlay removeFromSuperview];
            self.searchOverlay = nil;
        }];
    }
}

#pragma mark TableView

- (ISMSTagArtist *)tagArtistAtIndexPath:(NSIndexPath *)indexPath {
    if (self.isSearching && (self.dataModel.searchCount > 0 || self.searchBar.text.length > 0)) {
        return [self.dataModel tagArtistInSearchWithIndexPath:indexPath];
    } else {
        return [self.dataModel tagArtistWithIndexPath:indexPath];
    }
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView  {
    if (self.isSearching && (self.dataModel.searchCount > 0 || self.searchBar.text.length > 0)) {
        return 1;
    }
    return self.dataModel.tableSections.count;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    if (self.isSearching && (self.dataModel.searchCount > 0 || self.searchBar.text.length > 0)) {
        return self.dataModel.searchCount;
    } else if (self.dataModel.tableSections.count > section) {
        return self.dataModel.tableSections[section].itemCount;        
    }
    return 0;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    ISMSTagArtist *tagArtist = [self tagArtistAtIndexPath:indexPath];
    UniversalTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:UniversalTableViewCell.reuseId];
    [cell showCached:NO number:NO art:YES secondary:tagArtist.secondaryLabelText.hasValue duration:NO];
    [cell updateWithModel:[self tagArtistAtIndexPath:indexPath]];
    return cell;
}

- (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section {
    if (self.isSearching && (self.dataModel.searchCount > 0 || self.searchBar.text.length > 0)) return nil;
    if (section >= self.dataModel.tableSections.count) return nil;
    
    BlurredSectionHeader *sectionHeader = [tableView dequeueReusableHeaderFooterViewWithIdentifier:BlurredSectionHeader.reuseId];
    sectionHeader.text = self.dataModel.tableSections[section].name;
    return sectionHeader;
}

- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section {
    if (self.isSearching && (self.dataModel.searchCount > 0 || self.searchBar.text.length > 0)) return 0;
    if (section >= self.dataModel.tableSections.count) return 0;
    
    return Defines.rowHeight - 5;
}

- (NSArray *)sectionIndexTitlesForTableView:(UITableView *)tableView  {
    if (self.isSearching && (self.dataModel.searchCount > 0 || self.searchBar.text.length > 0)) return nil;
    
    NSMutableArray *titles = [NSMutableArray arrayWithCapacity:0];
    [titles addObject:@"{search}"];
    for (TableSection *section in self.dataModel.tableSections) {
        [titles addObject:section.name];
    }
    return titles;
}

- (NSInteger)tableView:(UITableView *)tableView sectionForSectionIndexTitle:(NSString *)title atIndex:(NSInteger)index  {
    if (self.isSearching && (self.dataModel.searchCount > 0 || self.searchBar.text.length > 0)) return -1;
    
    if (index == 0) {
        CGFloat yOffset = self.dropdown.hasMultipleMediaFolders ? 54 : 104;
        [self.tableView setContentOffset:CGPointMake(0, yOffset) animated:NO];
        return -1;
    }
    return index - 1;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath  {
    if (!indexPath) return;
    [self pushViewControllerCustom:[[TagArtistViewController alloc] initWithTagArtist:[self tagArtistAtIndexPath:indexPath]]];
}

- (void)tableView:(UITableView *)tableView willDisplayCell:(UITableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (!self.isSearching || !self.dataModel.shouldContinueSearch || self.dataModel.searchCount == 0 || self.searchBar.text.length == 0) return;
    
    // If we're searching and we've reached the last cell, continue the search
    if (indexPath.row == self.dataModel.searchCount - 1) {
        [EX2Dispatch runInBackgroundAsync:^{
            [self.dataModel continueSearch];
            [EX2Dispatch runInMainThreadAsync:^{
                [tableView reloadData];
            }];
        }];
    }
}

- (UISwipeActionsConfiguration *)tableView:(UITableView *)tableView trailingSwipeActionsConfigurationForRowAtIndexPath:(NSIndexPath *)indexPath {
    NSObject<TableCellModel> *model = [self tagArtistAtIndexPath:indexPath];
    return [SwipeAction downloadAndQueueConfigWithModel:model];
}

@end

File: Classes/View Controllers/Tabs/Folders/ArtistsViewController.swift
//
//  ArtistsViewController.swift
//  iSub
//
//  Created by Benjamin Baron on 1/15/21.
//  Copyright © 2021 Ben Baron. All rights reserved.
//

import UIKit
import Resolver
import SnapKit
import CocoaLumberjackSwift

@objc final class ArtistsViewController: UIViewController {
    @Injected private var store: Store
    @Injected private var settings: Settings
    
    private let tableView = UITableView()
    private lazy var dropdown = FolderDropdownControl(frame: CGRect(x: 50, y: 61, width: 220, height: 40))
    private let searchBar = UISearchBar()
    private var searchOverlay: UIVisualEffectView?
    private let countLabel = UILabel()
    private let reloadTimeLabel = UILabel()
    
    private var isSearching = false
    private var isCountShowing = false
    
    private let dataModel: ArtistsViewModel
        
    // MARK: Lifecycle
    
    init(dataModel: ArtistsViewModel) {
        self.dataModel = dataModel
        super.init(nibName: nil, bundle: nil)
    }
    
    required init?(coder: NSCoder) {
        fatalError("unimplemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = Colors.background
        title = "Folders"
        
        dataModel.delegate = self
        dataModel.reset()
        
        setupDefaultTableView(tableView)
        tableView.register(BlurredSectionHeader.self, forHeaderFooterViewReuseIdentifier: BlurredSectionHeader.reuseId)
        tableView.refreshControl = RefreshControl(handler: { [unowned self] in
            loadData(mediaFolderId: settings.currentServerId)
        })
        
        if dataModel.isCached {
            addCount()
            tableView.setContentOffset(CGPoint.zero, animated: false)
        }
        
        NotificationCenter.addObserverOnMainThread(self, selector: #selector(serverSwitched), name: ISMSNotification_ServerSwitched)
        NotificationCenter.addObserverOnMainThread(self, selector: #selector(updateFolders), name: ISMSNotification_ServerCheckPassed)
        NotificationCenter.addObserverOnMainThread(self, selector: #selector(addURLRefBackButton), name: UIApplication.didBecomeActiveNotification.rawValue)
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        addURLRefBackButton()
        addShowPlayerButton()
        if !dataModel.isCached {
            loadData(mediaFolderId: settings.rootFoldersSelectedFolderId?.intValue ?? 0)
        }
        Flurry.logEvent("FoldersTab")
    }
    
    deinit {
        NotificationCenter.removeObserverOnMainThread(self)
        dataModel.delegate = nil
        dropdown.delegate = nil
    }
    
    // MARK: Loading
    
    private func updateCount() {
        let count = dataModel.count
        countLabel.text = "\(count) \("Folder".pluralize(amount: count))"
        if let reloadDate = dataModel.reloadDate {
            let formatter = DateFormatter()
            formatter.dateStyle = .medium
            formatter.timeStyle = .short
            let formattedTime = formatter.string(from: reloadDate)
            reloadTimeLabel.text = "last reload \(formattedTime)"
        } else {
            reloadTimeLabel.text = ""
        }
    }
    
    private func removeCount() {
        tableView.tableHeaderView = nil
        isCountShowing = false
    }
    
    private func addCount() {
        isCountShowing = true
        
        let headerView = UIView()
        headerView.frame = CGRect(x: 0, y: 0, width: 320, height: 157)
        headerView.autoresizingMask = .flexibleWidth
        headerView.backgroundColor = view.backgroundColor

        countLabel.frame = CGRect(x: 0, y: 9, width: 320, height: 30)
        countLabel.autoresizingMask = .flexibleWidth
        countLabel.textColor = .label
        countLabel.textAlignment = .center
        countLabel.font = .boldSystemFont(ofSize: 32)
        headerView.addSubview(countLabel)
        
        reloadTimeLabel.frame = CGRect(x: 0, y: 40, width: 320, height: 14)
        reloadTimeLabel.autoresizingMask = .flexibleWidth
        reloadTimeLabel.textColor = .secondaryLabel
        reloadTimeLabel.textAlignment = .center
        reloadTimeLabel.font = .systemFont(ofSize: 11)
        headerView.addSubview(reloadTimeLabel)
        
        dropdown = FolderDropdownControl(frame: CGRect(x: 50, y: 61, width: 220, height: 40))
        dropdown.frame = CGRect(x: 50, y: 61, width: 220, height: 40)
        dropdown.delegate = self
        dropdown.selectFolder(withId: dataModel.mediaFolderId)
        headerView.addSubview(dropdown)
        // TODO: Why is this hack needed in the Swift port but not in the original Obj-C version??
        DispatchQueue.main.async(after: 0.1) {
            self.dropdown.frame = CGRect(x: 50, y: 61, width: self.view.frame.width - 100, height: 40)
            print(self.dropdown.frame)
        }


        searchBar.frame = CGRect(x: 0, y: 111, width: 320, height: 40)
        searchBar.autoresizingMask = .flexibleWidth
        searchBar.searchBarStyle = .minimal
        searchBar.delegate = self
        searchBar.autocorrectionType = .no
        searchBar.placeholder = "Folder name"
        headerView.addSubview(searchBar)

        updateCount()
        
        // Special handling for voice over users
        if UIAccessibility.isVoiceOverRunning {
            // Add a refresh button
            let voiceOverRefresh = UIButton(type: .custom)
            voiceOverRefresh.frame = CGRect(x: 0, y: 0, width: 50, height: 50)
            voiceOverRefresh.addTarget(self, action: #selector(reloadAction), for: .touchUpInside)
            voiceOverRefresh.accessibilityLabel = "Reload Folders"
            headerView.addSubview(voiceOverRefresh)

            // Resize the two labels at the top so the refresh button can be pressed
            countLabel.frame = CGRect(x: 50, y: 5, width: 220, height: 30)
            reloadTimeLabel.frame = CGRect(x: 50, y: 36, width: 220, height: 12)
        }
        
        tableView.tableHeaderView = headerView
    }
    
    @objc private func reloadAction() {
        loadData(mediaFolderId: settings.rootFoldersSelectedFolderId?.intValue ?? 0)
    }

    private func loadData(mediaFolderId: Int) {
        dropdown.updateFolders()
        HUD.show(closeHandler: cancelLoad)
        dataModel.mediaFolderId = mediaFolderId
        dataModel.startLoad()
    }
}

extension ArtistsViewController: APILoaderDelegate {
    func cancelLoad() {
        dataModel.cancelLoad()
        tableView.refreshControl?.endRefreshing()
        HUD.hide()
    }
    
    func loadingFinished(loader: APILoader?) {
        if isCountShowing {
            updateCount()
        } else {
            addCount()
        }
        
        tableView.reloadData()
        HUD.hide()
        tableView.refreshControl?.endRefreshing()
    }
    
    func loadingFailed(loader: APILoader?, error: NSError?) {
        HUD.hide()
        tableView.refreshControl?.endRefreshing()
        
        // Inform the user that the connection failed.
        // NOTE: Must call after a delay or the refresh control won't hide
        DispatchQueue.main.async(after: 0.3) {
            let alert = UIAlertController(title: "Subsonic Error", message: error?.localizedDescription ?? "Unknown error", preferredStyle: .alert)
            alert.addCancelAction(title: "OK")
            self.present(alert, animated: true, completion: nil)
        }
    }
}

extension ArtistsViewController: FolderDropdownDelegate {
    func folderDropdownMoveViewsY(_ y: Float) {
        tableView.performBatchUpdates({
            tableView.tableHeaderView?.frame.size.height += CGFloat(y)
            searchBar.frame.origin.y += CGFloat(y)
            tableView.tableHeaderView = tableView.tableHeaderView
            
            let indexPathsForVisibleRows = tableView.indexPathsForVisibleRows ?? []
            let visibleSections = Set<Int>(indexPathsForVisibleRows.map({ $0.section }))
            for section in visibleSections {
                if let sectionHeader = tableView.headerView(forSection: section) {
                    sectionHeader.frame.origin.y += CGFloat(y)
                }
            }
        }, completion: nil)
    }
    
    func folderDropdownSelectFolder(_ mediaFolderId: Int) {
        // Save the default
        settings.rootFoldersSelectedFolderId = NSNumber(value: mediaFolderId)
        
        // Reload the data
        dataModel.mediaFolderId = mediaFolderId
        isSearching = false
        if dataModel.isCached {
            tableView.reloadData()
            updateCount()
        } else {
            loadData(mediaFolderId: mediaFolderId)
        }
    }
    
    func folderDropdownViewsFinishedMoving() {
        
    }
    
    @objc private func serverSwitched() {
        dataModel.reset()
        if !dataModel.isCached {
            tableView.reloadData()
            removeCount()
        }
        folderDropdownSelectFolder(MediaFolder.allFoldersId)
    }
    
    @objc private func updateFolders() {
        dropdown.updateFolders()
    }
}

extension ArtistsViewController: UISearchBarDelegate {
    func searchBarTextDidBeginEditing(_ searchBar: UISearchBar) {
        guard !isSearching else { return }
        
        isSearching = true
        dataModel.clearSearch()
        
        dropdown.closeDropdownFast()
        tableView.setContentOffset(CGPoint(x: 0, y: 104), animated: true)
        
        if (searchBar.text?.count ?? 0) == 0 {
            createSearchOverlay()
        }
        
        // Add the done button
        navigationItem.leftBarButtonItem = UIBarButtonItem(barButtonSystemItem: .done, target: self, action: #selector(searchBarSearchButtonClicked(_:)))
    }
    
    func searchBarTextDidEndEditing(_ searchBar: UISearchBar) {
        searchBar.resignFirstResponder()
    }
    
    func searchBar(_ searchBar: UISearchBar, textDidChange searchText: String) {
        dataModel.clearSearch()
        if searchText.count > 0 {
            hideSearchOverlay()
            dataModel.search(name: searchText)
//            tableView.setContentOffset(CGPoint(x: 0, y: 45), animated: false)
        } else {
            createSearchOverlay()
//            tableView.setContentOffset(CGPoint(x: 0, y: 104), animated: false)
//            tableView.setContentOffset(CGPoint(x: 0, y: 45), animated: false)
        }
        tableView.reloadData()
        tableView.setContentOffset(CGPoint(x: 0, y: 45), animated: false)
    }
    
    @objc func searchBarSearchButtonClicked(_ searchBar: UISearchBar) {
        updateCount()
        
        self.searchBar.text = ""
        self.searchBar.resignFirstResponder()
        hideSearchOverlay()
        isSearching = false
        
        navigationItem.leftBarButtonItem = nil
        dataModel.clearSearch()
        tableView.reloadData()
        tableView.setContentOffset(CGPoint(x: 0, y: 104), animated: true)
    }
    
    private func createSearchOverlay() {
        let effectStyle: UIBlurEffect.Style = traitCollection.userInterfaceStyle == .dark ? .systemUltraThinMaterialLight : .systemUltraThinMaterialDark
        let searchOverlay = UIVisualEffectView(effect: UIBlurEffect(style: effectStyle))
        self.searchOverlay = searchOverlay
        view.addSubview(searchOverlay)
        searchOverlay.snp.makeConstraints { make in
            make.leading.trailing.bottom.equalToSuperview()
            make.top.equalToSuperview().offset(50)
        }
        
        let dismissButton = UIButton(type: .custom)
        dismissButton.addTarget(self, action: #selector(searchBarSearchButtonClicked(_:)), for: .touchUpInside)
        searchOverlay.contentView.addSubview(dismissButton)
        dismissButton.snp.makeConstraints { make in
            make.leading.trailing.top.bottom.equalToSuperview()
        }
        
        // Animate the search overlay on screen
        searchOverlay.alpha = 0
        UIView.animate(withDuration: 0.2, delay: 0, options: .curveEaseOut, animations: {
            searchOverlay.alpha = 1
        }, completion: nil)
    }
    
    private func hideSearchOverlay() {
        if let searchOverlay = searchOverlay {
            // Animate the search overlay off screen
            UIView.animate(withDuration: 0.2, delay: 0, options: .curveEaseOut) {
                searchOverlay.alpha = 0
            } completion: { _ in
                searchOverlay.removeFromSuperview()
                self.searchOverlay = nil
            }
        }
    }
}

extension ArtistsViewController: UITableViewDelegate, UITableViewDataSource {
    private func artist(indexPath: IndexPath) -> Artist? {
        if isSearching && (dataModel.searchCount > 0 || (searchBar.text?.count ?? 0) > 0) {
            return dataModel.artistInSearch(indexPath: indexPath)
        } else {
            return dataModel.artist(indexPath: indexPath)
        }
    }
    
    func numberOfSections(in tableView: UITableView) -> Int {
        if isSearching && (dataModel.searchCount > 0 || (searchBar.text?.count ?? 0) > 0) {
            return 1
        }
        return dataModel.tableSections.count
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if isSearching && (dataModel.searchCount > 0 || (searchBar.text?.count ?? 0) > 0) {
            return dataModel.searchCount
        } else if section < dataModel.tableSections.count {
            return dataModel.tableSections[section].itemCount
        }
        return 0
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueUniversalCell()
        cell.update(model: artist(indexPath: indexPath))
        return cell
    }
    
    func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        if isSearching && (dataModel.searchCount > 0 || (searchBar.text?.count ?? 0) > 0) { return nil }
        if section >= dataModel.tableSections.count { return nil }
        
        let sectionHeader = tableView.dequeueReusableHeaderFooterView(withIdentifier: BlurredSectionHeader.reuseId) as! BlurredSectionHeader
        sectionHeader.text = dataModel.tableSections[section].name
        return sectionHeader
    }
    
    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        if isSearching && (dataModel.searchCount > 0 || (searchBar.text?.count ?? 0) > 0) { return 0 }
        if section >= dataModel.tableSections.count { return 0 }
        
        return Defines.rowHeight - 5
    }
    
    func sectionIndexTitles(for tableView: UITableView) -> [String]? {
        if isSearching && (dataModel.searchCount > 0 || (searchBar.text?.count ?? 0) > 0) { return nil }
        
        var titles = ["{search}"]
        for section in dataModel.tableSections {
            titles.append(section.name)
        }
        return titles
    }
    
    func tableView(_ tableView: UITableView, sectionForSectionIndexTitle title: String, at index: Int) -> Int {
        if isSearching && (dataModel.searchCount > 0 || (searchBar.text?.count ?? 0) > 0) { return -1 }
        
        if index == 0 {
            let yOffset: CGFloat = dropdown.hasMultipleMediaFolders() ? 54 : 104
            tableView.setContentOffset(CGPoint(x: 0, y: yOffset), animated: false)
            return -1
        }
        return index - 1
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        if let artist = artist(indexPath: indexPath) {
            if let artist = artist as? FolderArtist {
                pushViewControllerCustom(FolderAlbumViewController(folderArtist: artist))
            } else if let artist = artist as? TagArtist {
                pushViewControllerCustom(TagArtistViewController(tagArtist: artist))
            }
        }
    }
    
    func tableView(_ tableView: UITableView, trailingSwipeActionsConfigurationForRowAt indexPath: IndexPath) -> UISwipeActionsConfiguration? {
        return SwipeAction.downloadAndQueueConfig(model: artist(indexPath: indexPath))
    }
}
